core/genesis.go

```diff
@@ -180,7 +180,18 @@ func SetupGenesisBlockWithOverride(db ethdb.Database, genesis *Genesis, override
 	// We have the genesis block in database(perhaps in ancient database)
 	// but the corresponding state is missing.
 	header := rawdb.ReadHeader(db, stored, 0)
-	if _, err := state.New(header.Root, state.NewDatabaseWithConfig(db, nil), nil); err != nil {
+
+	var trieCfg *trie.Config
+
+	if genesis == nil {
+		storedcfg := rawdb.ReadChainConfig(db, stored)
```

Comment:
why not create a `DefaultGenesisBlock` here? based on the logic below, it seems to allow `genesis` to be nil here.

Reply:
I think the `DefaultGenesisBlock` is created in the code blow (line 195). The if closure here tends to read the zktrie flag for stateDB in case of genesis is not avaliable yet. In the original code, stateDB can be created with default configuration but now we had to know at least the zktrie flag beforehand.

Reply:
I feel it's bad to panic here. An alternative way is that we just set `zktrie` to be true by default.

Reply:
I think we had to set zktrie to be false in this case because:

1. If no config can be read here and no root state has been saved into stateDb, the genesis and configuration would be initialized to mainnet's.
2. If there has been a root state being saved and no config can be read, it would be initialized later by some default configurations (in line 220) and act as the new chain config data (line 228).

All of there configruations above has no zktrie being enabled. Anyway, I think it would be a better practice to only active zktrie flag explicitly by the genesis object and left other cases unchanged.

Reply:
Makes sense. One quick question. Do you allow `trieCfg` to be nil in line 197 since `trieCfg` could be nil if you enter line 189?

Reply:
Yes. In the original code they just pass a nil config. Use nil trieCfg just means respect the original behavior.